File: 1_scripts/libs/pipeline.py
# -*- coding: utf-8 -*-
"""
pipeline.py
============
Implements the folder convention from Ties de Kok, "How to keep your
research projects organized, part 1: folder structure" (see this repo's
top-level README.md). Each script in 1_scripts/ gets its own pipeline
sub-folder (name-matched) under 2_pipeline/:

    out/   end-products meant to be loaded by a LATER script
    store/ intermediate results cached for reloading by THIS SAME script
    tmp/   throwaway inspection files, safe to delete anytime

`get_pipeline_paths(__file__)` derives both the script's name and its
project root from the script's own file path -- call it as the first line
of any script living directly in `1_scripts/` and it resolves
`2_pipeline/<script_name>/` automatically, creating it if missing.

NOTE: this assumes scripts live directly in `1_scripts/` (one level below
the project root) -- i.e. `<root>/1_scripts/x.py`, so `x.py`'s parent's
parent is `<root>`. If you ever nest scripts deeper, this needs a matching
depth change (see the R version, pipeline.R, for the equivalent).
"""

from dataclasses import dataclass
from pathlib import Path


@dataclass
class PipelinePaths:
    out: Path
    store: Path
    tmp: Path


def get_pipeline_paths(script_file, subdirs=("out", "store", "tmp")):
    script_path = Path(script_file).resolve()
    name = script_path.stem                # e.g. "1_behavior"
    root = script_path.parent.parent        # <root>/1_scripts/x.py -> <root>/
    pipeline_dir = root / "2_pipeline" / name

    paths = {}
    for sub in subdirs:
        p = pipeline_dir / sub
        p.mkdir(parents=True, exist_ok=True)
        paths[sub] = p

    if subdirs == ("out", "store", "tmp"):
        return PipelinePaths(out=paths["out"], store=paths["store"], tmp=paths["tmp"])
    return paths


def project_root(script_file, marker="pyproject.toml"):
    """Walk up from script_file until `marker` is found. Depth-agnostic
    (loop-based), unlike get_pipeline_paths -- use this whenever you need
    the true project root (e.g. to reach 0_data/)."""
    p = Path(script_file).resolve().parent
    while p != p.parent:
        if (p / marker).exists():
            return p
        p = p.parent
    raise FileNotFoundError(f"Could not find {marker} walking up from {script_file}")

File: 1_scripts/libs/pipeline.R

# ==============================================================================
# pipeline.R
# ==============================================================================
# R equivalent of infotheory/pipeline.py -- same out/store/tmp convention.
# Source this at the top of any script living directly in `1_scripts/` --
# see any of 1_behavior.R / 2_single.R for the exact one-line bootstrap to
# copy (it has to locate ITSELF relative to the running script first).
#
# NOTE: depth-dependent, same caveat as pipeline.py -- assumes scripts live
# directly in `1_scripts/` (one level below project root).
# ==============================================================================

# ==============================================================================
# pipeline.R
# ==============================================================================
# R has no reliable equivalent of Python's `__file__` -- one that works
# identically under `Rscript`, `source()`, and interactive/RStudio
# execution. Rather than keep patching heuristics for each invocation mode
# (which is what the previous version of this file did, and which still
# broke under `source()`/RStudio's Source button), this version follows
# the source article's own advice instead: always run with the WORKING
# DIRECTORY set to the project root, and use relative paths from there.
#
# Practically: open this repo as an RStudio Project (File -> Open Project,
# or double-click a .Rproj file at the repo root) and RStudio sets the
# working directory correctly every time, automatically. From a terminal,
# `cd` to the repo root before running `Rscript 1_scripts/1_behavior.R`.
#
# The tradeoff for this robustness: each script needs one extra line, a
# NAME constant -- matching the original de Kok generator's own template
# (see 1_scripts/1_behavior.R for the exact pattern to copy).
# ==============================================================================

check_project_root <- function() {
  ok <- dir.exists("1_scripts") && dir.exists("0_data")
  if (!ok) {
    stop(
      "Working directory doesn't look like the project root (expected to find ",
      "1_scripts/ and 0_data/ here). Current working directory: ", getwd(), "
",
      "Fix: setwd() to the repo root before running this script, or open the ",
      "repo as an RStudio Project so the working directory is set automatically."
    )
  }
  invisible(TRUE)
}

get_pipeline_paths <- function(name, subdirs = c("out", "store", "tmp")) {
  check_project_root()
  pipeline_dir <- file.path("2_pipeline", name)
  paths <- list()
  for (sub in subdirs) {
    p <- file.path(pipeline_dir, sub)
    dir.create(p, showWarnings = FALSE, recursive = TRUE)
    paths[[sub]] <- p
  }
  paths
}

project_root <- function() {
  check_project_root()
  normalizePath(getwd())
}
